### PR TITLE
js-git-ipld <-> pando library

### LINKED ISSUE #50: pando.js
Hey,

Here is a summary of the discussion we've had yesterday with Cory regarding the implementation of a small `pando.js` library.

## What

The goal of this library would be to:

- provide a more generic and efficient way to handle repositories frontend state
- offer in-browser markdown edition,

## How 

To do so, we need to implement a small library providing the following features:

- take a head commit IPLD CID and reduce it into an historically ordered commit array [the commit history of the given branch] as done in the current background script [here](https://github.com/pandonetwork/pando/blob/344171694b4cac36f48ee6694450c812191c646a/packages/pando-repository/app/src/script.js#L40)
- turn an IPLD linked-data commit object into an easier to manipulate Javascript / Typescript object
- extract each file last update date and message for these commits
- take a Javascript / Typescript object and turn it back into an IPLD / linked data commit object [to easily create a new commit to push from the browser]

## Details

`js-git-ipld` commit objects look like that:

```
{
  "author": {
    "date": "1503667703 +0200",
    "email": "author@mail",
    "name": "Author Name"
  },
  "committer": {
    "date": "1503667703 +0200",
    "email": "author@mail",
    "name": "Author Name"
  },
  "message": "Commit Message\n",
  "parents": [
    { "/": cid }, { "/": cid }, ...
  ],
  "tree": { "/", cid }
}
```

The idea would be for this library to take the CID of such a given commit [the HEAD cid of a branch] and turn it into the following ordered array:

```
[
  commitN,
  commitN-1,
  commitN-2,
  ...
]
```

Where each commit would look like that:

```
{
  "cid": zw9X ..., // CID of the commit
  "sha": 9c4v ..., // git sha1 of the commit
  " "author": {
    "date": "1503667703 +0200",
    "email": "author@mail",
    "name": "Author Name"
  },
  "committer": {
    "date": "1503667703 +0200",
    "email": "author@mail",
    "name": "Author Name"
  },
  "message": "Commit Message\n",
  "tree": {
    "README.md": { "lastEdit": { "date": xxx, "message": xxx }, mode: "12600", "blob":  zw9x ... // CID of the file },
    "app/index.js": { "lastEdit": { "date": xxx, "message": xxx }, mode: "12600", "blob":  zw9x ... // CID of the file  },
    "app/util/index.js": { "lastEdit": { "date": xxx, "message": xxx }, mode: "12600", "blob":  zw9x ... // CID of the file },
     ...
  },
  parents: [
    zw9x ..., // CID of parent commit
    zw9x ... // CID of the
    ...
  ]
}
```

Conversely, this library should be able to take the object described below as an input an turn it back into an IPLD commit.

### PR DESCRIPTION
**Pull request by corydickson**
Looking for some feedback on this approach before I deprecate IPLD functions inside of `script.js`

Todo:
- [ ] Convert from `Commit` class back into IPLD format.
- [ ] Use this lib for state management

Resolves #50

### CODE REVIEW

File: packages/pando-repository/app/src/lib/pando.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        this.sha = cidToSha(commit.tree['/']).toString('hex')
+        this.author = commit.author
+        this.committer = commit.committer
+        this.message = commit.message
+        this.tree = commit.tree
+        this.parents = _parents
+    }
+  }
+
+  public async fetchModifiedTree(): Promise<void> {
```
Comment (corydickson) [inline packages/pando-repository/app/src/lib/pando.ts]: I came upon this solution given that we couldn't traverse the `tree` inside the `constructor`

File: packages/pando-repository/app/src/script.js [hunk 2]
```diff
@@ -156,3 +141,25 @@ const fetchFilesFromTree = (hash, path) => {
     })
   })
 }
+
+const fetchOrderedHistory = async (hash, formerHistory) => {
```
Comment (corydickson) [inline packages/pando-repository/app/src/script.js:146]: In the final version of this PR, we'll deprecate all of these helper functions in this file and import them from `pando.ts`

File: packages/pando-repository/app/src/lib/pando.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  date: string
+}
+
+export interface ITree { [path: string]: string }
+
+export interface ILinkedDataCommit {
+  author: IAuthorOrCommitter
+  committer: IAuthorOrCommitter
+  message: string
+  parents: ITree[] | null
```
Comment (osarrouy) [inline packages/pando-repository/app/src/lib/pando.ts]: Isn't it supposed to be : `parents: ILinkedCommitData[]` ?

**Review by osarrouy — COMMENTED** [review]
Hey ! I added a couple of comment. Also I think the best would be to: 
1. split this lib into its own package. We may need to use it in other places [especially in the GraphProtocol reducer if we come up with one]
2. Write test to make sure of the consistency of the lib. The best would be extract a known IPLD commit history build with pando and, in the test, push it to IPFS and then try to fetch it again / update it. Otherwise i think it will be super tricky to test the behaviour of the lib each time we update it.

**Comment by osarrouy** [comment]
Closing for now since we need to figure out an answer to #50 before ... Will re-open later if needed ...

**Pull request closed without merging**